projects/Mallard/src/helpers/queries.ts

```diff
@@ -0,0 +1,249 @@
+        if (node === undefined) return
+        this._refresh(node, EMPTY_STR_SET)
+    }
+
+    private _getKey<Value, Variables>(
+        query: Query<Value, Variables>,
+        variables: Variables,
+    ) {
+        return query.id + ':' + JSON.stringify(variables)
+    }
+
+    private _getNode<Value, Variables>(
+        query: Query<Value, Variables>,
+        variables: Variables,
+    ): QueryNode {
+        const key = this._getKey(query, variables)
+        let node = this._graph.get(key)
+        if (node !== undefined) return node
+
+        node = {
+            key,
+            query: (query as unknown) as Query<unknown, unknown>,
+            promise: undefined,
+            value: undefined,
+            error: undefined,
+            variables,
+            deps: new Set(),
+            preds: new Set(),
+            listeners: [],
+        }
+        this._graph.set(key, node)
+        this._refresh(node, EMPTY_STR_SET)
+        return node
+    }
+
+    private _refresh(node: QueryNode, trail: Set<string>) {
+        if (trail.has(node.key))
+            throw new Error('circular dependency between queries')
+
+        for (const depKey of node.deps) {
+            const dep = this._graph.get(depKey)
+            if (dep === undefined || !dep.preds.has(node.key))
+                throw new Error('inconsistent graph')
+            dep.preds.delete(node.key)
+        }
+        node.deps.clear()
+
+        const resolve = (
+            query: Query<unknown, unknown>,
+            variables: unknown,
+        ) => {
+            const subNode = this._getNode(query, variables)
+            node.deps.add(subNode.key)
+            subNode.preds.add(node.key)
+            if (subNode.promise !== undefined) return subNode.promise
+            if (subNode.value !== undefined)
+                return Promise.resolve(subNode.value)
+            if (subNode.error !== undefined)
+                return Promise.reject(subNode.error)
+            throw new Error('inconsistent graph')
+        }
+
+        const prevValue = node.value
+        const promise = Promise.resolve().then(() =>
+            node.query.resolver(
+                node.variables,
+                (resolve as unknown) as LocalResolver,
+                prevValue,
+            ),
+        )
+        node.promise = promise
+        promise
+            .then(
+                value => {
+                    if (node.promise !== promise) return
+                    node.promise = undefined
+                    node.value = value
+                },
+                error => {
+                    if (node.promise !== promise) return
+                    node.promise = undefined
+                    node.error = error
+                },
+            )
+            .then(() => {
+                for (const listener of node.listeners) {
+                    listener()
+                }
+            })
+            .catch(error => {
+                setImmediate(() => {
```

Comment:
We can't throw directly in the `catch` clause because it would immediatly get handled by the `Promise` system, and transformed into an (unhandled) rejection. So we rethrow "in the void" so that we get caught by the global error handler (we need to check that we do have a global error handler, which we really should if not.)

Reply:
That was my point, I'm not sure we do have a global error handler at the process level, I might be wrong though.